db/src/tx_checking.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::{
+    BTreeSet,
+    BTreeMap,
+};
+
+use mentat_core::{
+    Entid,
+    TypedValue,
+    ValueType,
+};
+
+use errors::{
+    CardinalityConflict,
+};
+
+use internal_types::{
+    AEVTrie,
+};
+
+/// Map from found [e a v] to expected type.
+pub(crate) type TypeDisagreements = BTreeMap<(Entid, Entid, TypedValue), ValueType>;
+
+/// Ensure that the given terms type check.
+///
+/// We try to be maximally helpful by yielding every malformed datom, rather than only the first.
```

Comment:
I would rather see this done by receiving an error on insert of each datom and collecting the errors, continuing anyway. It won't catch _every_ error, but it'll catch most, and it's a streaming approach that fails early in the common case.